src/librustc/traits/mod.rs

```diff
@@ -242,6 +244,59 @@ pub type Obligations<'tcx, O> = Vec<Obligation<'tcx, O>>;
 pub type PredicateObligations<'tcx> = Vec<PredicateObligation<'tcx>>;
 pub type TraitObligations<'tcx> = Vec<TraitObligation<'tcx>>;
 
+#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug)]
+pub enum WhereClauseAtom<'tcx> {
+    Implemented(ty::PolyTraitPredicate<'tcx>),
```

Comment:
I don't think we want poly trait predicates here. I think we want the binders to be moved out of `WhereClauseAtom` and into the `Goal` -- actually, in Chalk we get this mildly wrong, as I'll comment below.